File: lib/src/articles/dtos/article_dto.dart
import 'package:dart_shelf_realworld_example_app/src/profiles/dtos/profile_dto.dart';

class ArticleDto {
  final String slug;
  final String title;
  final String description;
  final String body;
  final List<String> tagList;
  final DateTime createdAt;
  final DateTime updatedAt;
  final bool favorited;
  final int favoritesCount;
  final ProfileDto author;

  ArticleDto(
      {required this.slug,
      required this.title,
      required this.description,
      required this.body,
      required this.tagList,
      required this.createdAt,
      required this.updatedAt,
      required this.favorited,
      required this.favoritesCount,
      required this.author});

  ArticleDto.fromJson(Map<String, dynamic> json)
      : slug = json['article']['slug'],
        title = json['article']['title'],
        description = json['article']['description'],
        body = json['article']['body'],
        tagList = List<String>.from(json['article']['tagList']),
        createdAt = DateTime.parse(json['article']['createdAt']),
        updatedAt = DateTime.parse(json['article']['updatedAt']),
        favorited = json['article']['favorited'],
        favoritesCount = json['article']['favoritesCount'],
        author = ProfileDto.fromJson(json['article']['author']);

  Map<String, dynamic> toJson() => {
        'article': {
          'slug': slug,
          'title': title,
          'description': description,
          'body': body,
          'tagList': tagList,
          'createdAt': createdAt.toIso8601String(),
          'updatedAt': updatedAt.toIso8601String(),
          'favorited': favorited,
          'favoritesCount': favoritesCount,
          'author': author.toJson()
        }
      };
}

File: src/main/scala/com/softwaremill/realworld/articles/ArticleDto.scala
package com.softwaremill.realworld.articles

import com.softwaremill.realworld.common.NoneAsNullOptionEncoder.*
import zio.json.{DeriveJsonDecoder, DeriveJsonEncoder}

import java.time.Instant

case class Article(
    article: ArticleData
)
object Article:
  given articleEncoder: zio.json.JsonEncoder[Article] = DeriveJsonEncoder.gen[Article]
  given articleDecoder: zio.json.JsonDecoder[Article] = DeriveJsonDecoder.gen[Article]

case class ArticleCreate(
    article: ArticleCreateData
)
object ArticleCreate:
  given articleCreateEncoder: zio.json.JsonEncoder[ArticleCreate] = DeriveJsonEncoder.gen[ArticleCreate]
  given articleCreateDecoder: zio.json.JsonDecoder[ArticleCreate] = DeriveJsonDecoder.gen[ArticleCreate]

case class ArticleUpdate(
    article: ArticleUpdateData
)
object ArticleUpdate:
  given articleUpdateEncoder: zio.json.JsonEncoder[ArticleUpdate] = DeriveJsonEncoder.gen[ArticleUpdate]
  given articleUpdateDecoder: zio.json.JsonDecoder[ArticleUpdate] = DeriveJsonDecoder.gen[ArticleUpdate]

case class ArticleData(
    slug: String,
    title: String,
    description: String,
    body: String,
    tagList: List[String],
    createdAt: Instant,
    updatedAt: Instant,
    favorited: Boolean,
    favoritesCount: Int,
    author: ArticleAuthor
)
object ArticleData:
  given articleDataEncoder: zio.json.JsonEncoder[ArticleData] = DeriveJsonEncoder.gen[ArticleData]
  given articleDataDecoder: zio.json.JsonDecoder[ArticleData] = DeriveJsonDecoder.gen[ArticleData]

case class ArticleCreateData(
    title: String,
    description: String,
    body: String,
    tagList: List[String] // TODO tagList is currently mandatory...
)

object ArticleCreateData:
  given articleCreateDataEncoder: zio.json.JsonEncoder[ArticleCreateData] = DeriveJsonEncoder.gen[ArticleCreateData]
  given articleCreateDataDecoder: zio.json.JsonDecoder[ArticleCreateData] = DeriveJsonDecoder.gen[ArticleCreateData]

case class ArticleUpdateData(
    slug: Option[String], // TODO add additional class without slug
    title: Option[String],
    description: Option[String],
    body: Option[String]
)
object ArticleUpdateData:
  given articleUpdateDataEncoder: zio.json.JsonEncoder[ArticleUpdateData] = DeriveJsonEncoder.gen[ArticleUpdateData]
  given articleUpdateDataDecoder: zio.json.JsonDecoder[ArticleUpdateData] = DeriveJsonDecoder.gen[ArticleUpdateData]

case class ArticleAuthor(
    username: String,
    bio: Option[String],
    image: Option[String],
    following: Boolean
)
object ArticleAuthor:
  given articleAuthorEncoder: zio.json.JsonEncoder[ArticleAuthor] = DeriveJsonEncoder.gen[ArticleAuthor]
  given articleAuthorDecoder: zio.json.JsonDecoder[ArticleAuthor] = DeriveJsonDecoder.gen[ArticleAuthor]

case class ArticleTagRow(
    tag: String,
    articleSlug: String
)

case class ArticleFavoriteRow(
    profileId: Int,
    articleSlug: String
)

case class ArticleRow(
    slug: String,
    title: String,
    description: String,
    body: String,
    createdAt: Instant,
    updatedAt: Instant,
    authorId: Int
)
